### ISSUE TITLE
Instrumentation hangs when the yield keyword is used in a property that's part of an explicit interface implementation

### ISSUE THREAD

**Issue by Anvoker** [SteveGilham/altcover#43]
I've encountered a really peculiar bug. If altcover is told to instrument a .dll that contains a property with a getter that returns an IEnumerable<T> and uses the yield keyword anywhere in its body AND that property is part of an explicit interface implementation then altcover will get permanently stuck in an infinite loop and never finish the instrumentation process. It will use up to 25% of my CPU and never give an error. The instrumented `AltCoverRepro2.dll` seems to be generated during this process, but the XML file, `AltCover.Recorder.g.dll`, and `AltCover.Recorder.g.pdb` are not generated which prevents me from using a runner.

This is the command I used:

```
H:\Programe\altcover\tools\net45\AltCover.exe --opencover -s=nunit -t=System. -s=Microsoft -i=F:\GitRepos\AltCoverRepro2\AltCoverRepro2\bin\Debug -o=F:\GitRepos\AltCoverRepro2\AltCoverRepro2\bin\Debug\instr -x=F:\GitRepos\AltCoverRepro2\AltCoverRepro2\bin\Debug\coverage.xml
```

As soon as the offending yield usage is stubbed out with a `throw new NotImplementedException()` then the exact same command will successfully instrument the project. The three previously missing files are successfully generated. The project is also instrumented successfully if the property is switched from being explicit to being public.

I have attached a repro project at the bottom but will include a reproduction inline here as well. It's also worth noting that I'm on Windows 10 Pro and this is being compiled with .NET Framework / Roslyn. I opted for this mode of operation because I wanted to be doing similar steps to what I'd do on Linux with Mono, since I'd like to move to developing on Linux.

```csharp
public class EnumerableGetterExplicit_DoesNotWork<T, K>
    : IReadOnlyDictionary<T, K>
{
    private int counter;

    // Replace the content of this property's getter with
    // "=> throw new NotImplementedException();"
    // and the whole project is instrumented successfully.
    // As long as this getter is left alone, instrumentation enters an
    // infinite loop.
    IEnumerable<K> IReadOnlyDictionary<T, K>.Values
    {
        get { yield break; }
    }

    // This works as is. It seems the problem doesn't appear unless the
    // property is specifically part of an explicit interface
    // implementation.
    public IEnumerable<K> ValuesWorks
    {
        get
        {
            yield return default(K);
            yield break;
        }
    }

    // This works as is. As long as there isn't a yield used inside this
    // getter then the property won't cause any trouble to the
    // instrumentation process.
    IEnumerable<T> IReadOnlyDictionary<T, K>.Keys
    {
        get { counter++; throw new NotImplementedException(); }
        // The counter variable is just a blind hedge against the compiler
        // doing something weird. It might optimize the getter out somehow
        // if the body of the getter was just the exception throw.
    }

    K IReadOnlyDictionary<T, K>.this[T key] => throw new NotImplementedException();
    int IReadOnlyCollection<KeyValuePair<T, K>>.Count => throw new NotImplementedException();
    bool IReadOnlyDictionary<T, K>.ContainsKey(T key) => throw new NotImplementedException();
    IEnumerator<KeyValuePair<T, K>> IEnumerable<KeyValuePair<T, K>>.GetEnumerator() => throw new NotImplementedException();
    IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
    bool IReadOnlyDictionary<T, K>.TryGetValue(T key, out K value) => throw new NotImplementedException();
}
```
[AltCoverRepro_ExplicitGetterYieldInfiniteLoop2.zip](https://github.com/SteveGilham/altcover/files/2665262/AltCoverRepro_ExplicitGetterYieldInfiniteLoop2.zip)

I've struggled literally all day to get this work until I managed to condense it down to this one bizarre case. My actual project is much larger than this but I managed to confirm this is the problem by stubbing out all occurrences of the keyword yield in explicit interface members - this made altcover successfully instrument the project right away. Usage of strong name keys did not seem to matter. My guess is that whatever code generates the IL necessary for instrumentation is somehow failing to parse this specific case.

**Comment by SteveGilham** [comment]
Running this is the debugger, the loop turns out to be in the code that attempts to discover which user-defined class/method the various compiler generated classes/methods are associated with (so as to be able to correctly apply exclusions).  Method `System.Collections.Generic.IEnumerator`1<K> AltCoverRepro.EnumerableGetterExplicit_DoesNotWork`2/<System-Collections-Generic-IReadOnlyDictionary<T,K>-get_Values>d__2::System.Collections.Generic.IEnumerable<K>.GetEnumerator()` is being detected as containing itself.

Bug looks to be that I'm not correctly matching the angle-brackets correctly for names like `<System-Collections-Generic-IReadOnlyDictionary<T,K>-get_Values>`

**Comment by SteveGilham** [comment]
Actually two bugs; the second one -- the one that really led to the looping -- being that the `GetEnumerator()` of that synthetic type recursively constructs instances of itself.  Simple recursion like this is easy to detect.  If there is ever mutual recursion amongst the synthetic types then more work will be required, but I shall declare YAGNI and wait to see an example in the wild before coding that.

Look for a bugfix release in the next few days.

**Comment by SteveGilham** [comment]
Should be fixed in release 4.0.660

**Comment by Anvoker** [comment]
Tried the new build. Instrumentation finished and generated all the expected files. Thank you, the fixed is much appreciated!
